### PR TITLE
[Pavlo Turchyn]UIT-44 related products initial commit

### PR DESCRIPTION
**Pull request by FabianMoretti**

### CODE REVIEW

File: src/app/product-details-page/components/product-details-page/product-details-page.component.spec.ts [hunk 1]
```diff
@@ -23,7 +25,7 @@ describe('ProductDetailsPageComponent', () => {
     fixture.detectChanges();
   });
 
-  it('should create', () => {
+  xit('should create', () => {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.spec.ts]: why did you skip it?

File: src/app/product-details-page/product-details-page.module.ts [hunk 2]
```diff
@@ -5,12 +5,15 @@ import { ProductDetailsPageRoutingModule } from './product-details-page-routing.
 import { ProductDescriptionComponent } from './components/product-description/product-description.component';
 import { ProductDetailsPageComponent } from './components/product-details-page/product-details-page.component';
 import { ProductImagePreviewComponent } from './components/product-description/product-image-preview/product-image-preview.component';
+import { SharedModule } from '../shared/shared.module';
+import { HeadingComponent } from '../shared/components/heading/heading.component';
 
 @NgModule({
   declarations: [ProductDetailsPageComponent, ProductDescriptionComponent, ProductImagePreviewComponent],
   imports: [
     CommonModule,
```
Comment (dverbovyi) [inline src/app/product-details-page/product-details-page.module.ts]: remove CommonModule. It's already in the shared

File: src/app/shared/components/related-products/related-products.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(RelatedProductsComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  xit('should create', () => {
```
Comment (dverbovyi) [inline src/app/shared/components/related-products/related-products.component.spec.ts]: why did you skip it?

File: src/app/shared/shared.module.ts [hunk 4]
```diff
@@ -14,6 +14,7 @@ import { JoinUsComponent } from './components/join-us/join-us.component';
 import { AdvertisingAreaComponent } from './components/advertising-area/advertising-area.component';
 import { SafePipe } from '../pipes/safe.pipe';
 import { HeadingComponent } from './components/heading/heading.component';
+import { RelatedProductsComponent } from './components/related-products/related-products.component';
```
Comment (dverbovyi) [inline src/app/shared/shared.module.ts]: This is not shared component. It's only related to PDP

File: src/app/shared/shared.module.ts [hunk 5]
```diff
@@ -33,7 +34,8 @@ const MODULES = [
     JoinUsComponent,
     ProductItemComponent,
     ProductItemShortComponent,
-    HeadingComponent
+    HeadingComponent,
+    RelatedProductsComponent
```
Comment (dverbovyi) [inline src/app/shared/shared.module.ts]: This is not shared component. It's only related to PDP

File: src/app/shared/shared.module.ts [hunk 6]
```diff
@@ -45,7 +47,8 @@ const MODULES = [
...
```
Comment (dverbovyi) [inline src/app/shared/shared.module.ts]: RelatedProductsComponent is not shared component. It's only related to PDP

**Review by dverbovyi — CHANGES_REQUESTED** [review]

**Comment by dverbovyi** [comment]
I provided a technical vision in story comment

https://issues.griddynamics.net/browse/UIT-44

**Comment by FabianMoretti** [comment]
https://issues.griddynamics.net/browse/UIT-44

File: src/app/home-page/components/home-page/home-page.component.ts [hunk 7]
```diff
@@ -22,6 +22,7 @@ export class HomePageComponent {
     sizes: [ProductSize.S, ProductSize.M, ProductSize.L, ProductSize.XL],
     addedToCart: false,
     addedToWishList: true,
+    relatedProducts: [],
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.ts]: we don't need related products for Home-Page

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import { ActivatedRoute } from '@angular/router';
+declare var require: any;
+
+@Component({
+  selector: 'app-related-products',
+  templateUrl: './related-products.component.html',
+  styleUrls: ['./related-products.component.scss']
+})
+export class RelatedProductsComponent implements OnInit {
+  json = require('../../../../mocks/pdp.json');
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: it should be HTTP request. Create PDPService which consumes DataService as Transport

File: src/app/mocks/pdp.json [hunk 9]
```diff
@@ -0,0 +1,318 @@
+[{
```
Comment (dverbovyi) [inline src/app/mocks/pdp.json]: it shouldn't be an array. Only 1 product with related inside

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.html [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<div *ngIf="this.relatedProductsArr.length > 0">
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.html]: > 0 can be omitted,  

just.length

File: src/app/shared/interfaces/product.ts [hunk 11]
```diff
@@ -9,6 +9,7 @@ export interface IProduct {
   sizes?: Array<ProductSize>;
   addedToCart: Boolean;
   addedToWishList: Boolean;
+  relatedProducts: Array<IProduct>;
```
Comment (dverbovyi) [inline src/app/shared/interfaces/product.ts]: this is optional 

relatedProducts?: Array<IProduct>

**Review by dverbovyi — CHANGES_REQUESTED** [review]
@FabianMoretti 

related-products.component.scss

File: src/app/home-page/components/home-page/home-page.component.ts [hunk 12]
```diff
@@ -35,6 +36,7 @@ export class HomePageComponent {
     sizes: [ProductSize.S, ProductSize.M, ProductSize.L, ProductSize.XL],
     addedToCart: false,
     addedToWishList: false,
+    relatedProducts: [],
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.ts]: delete related from home-page

File: src/app/core/services/product-details-page.service.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductDetailsPageService {
+
+
+  constructor(private mocks: DataService) {}
+
+  getRelatedProducts() {
```
Comment (dverbovyi) [inline src/app/core/services/product-details-page.service.ts]: getProduct(id:String)

File: src/app/core/services/product-details-page.service.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+import { DataService } from './data.service';
+import { BehaviorSubject } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductDetailsPageService {
+
+
+  constructor(private mocks: DataService) {}
```
Comment (dverbovyi) [inline src/app/core/services/product-details-page.service.ts]: private dataService: DataService

File: src/app/product-details-page/components/product-details-page/product-details-page.component.html [hunk 15]
```diff
@@ -1 +1,2 @@
 <app-product-description [product]="productInfo"></app-product-description>
+<app-related-products></app-related-products>
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.html]: related products should be passed as an attribute from product-details component
<app-related-products [products]="relatedProducts" ></app-related-products>

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.html [hunk 16]
```diff
@@ -0,0 +1,8 @@
+<div *ngIf="relatedProductsArrLength">
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.html]: relatedProducts.length

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+import { IProduct } from 'src/app/shared/interfaces/product';
+import { ProductDetailsPageService } from 'src/app/core/services/product-details-page.service';
+
+@Component({
+  selector: 'app-related-products',
+  templateUrl: './related-products.component.html'
+})
+export class RelatedProductsComponent implements OnInit {
+
+  products: IProduct;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: Input() products: IProduct[] = [];

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  id: number;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: delete the all stuff below

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+  productsArr;
+  relatedProductsArr;
+  relatedProductsArrLength;
+  test: number;
+
+
+  constructor(private mock: ProductDetailsPageService) { }
+
+  ngOnInit() {
+    this.mock.getRelatedProducts().subscribe(data => {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: the request should happened in the page-level component.

you should 

```
this.dataService.getProduct(routeId)
```

File: src/app/shared/pipes/safe.pipe.spec.ts [hunk 20]
```diff
@@ -1,7 +1,7 @@
 import { SafePipe } from './safe.pipe';
 import { TestBed } from '@angular/core/testing';
 import { DomSanitizer } from '@angular/platform-browser';
-import { AdvertisingAreaComponent } from '../shared/components/advertising-area/advertising-area.component';
+import { AdvertisingAreaComponent } from '../components/advertising-area/advertising-area.component';
```
Comment (dverbovyi) [inline src/app/shared/pipes/safe.pipe.spec.ts]: This test related to Pipe. no components should be here

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/core/services/product-details-page.service.spec.ts [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+
+describe('ProductDetailsPageService', () => {
+  beforeEach(() => TestBed.configureTestingModule({
+    imports: [
+      HttpClientModule,
+      HttpClientTestingModule
+    ]
+  }));
+
+  it('should be created', () => {
```
Comment (dverbovyi) [inline src/app/core/services/product-details-page.service.spec.ts]: add an additional test for the service which validates getProduct

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
+import { Component, Input, OnChanges } from '@angular/core';
+import { IProduct } from 'src/app/shared/interfaces/product';
+
+@Component({
+  selector: 'app-related-products',
+  templateUrl: './related-products.component.html'
+})
+export class RelatedProductsComponent implements OnChanges {
+
+  @Input() products: IProduct[];
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: @Input() products: IProduct[] = []; //define with a default value

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+
+  ngOnChanges(): void {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: you don't need this

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 24]
```diff
@@ -24,4 +27,13 @@ export class ProductDetailsPageComponent {
...
     description: 'Lorem Lorem Lorem',
   };
+  routeId: String;
+  relatedProducts: IProduct[] = [];
+
+  constructor(private route: ActivatedRoute, private dataService: ProductDetailsPageService) { }
+
+  ngOnInit() {
+    this.route.params.subscribe(data => this.routeId = data.id);
+    this.dataService.getProduct(this.routeId).subscribe(data => this.relatedProducts = data.relatedProducts);
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: on this line the routeId is undefined

use the next

```ts
ngOnInit() {
   this.productSubscription = this.route.params.pipe(
      mergeMap(({id)) => this.dataService.getProduct(id)
   ).subscribe(product => this.product = product);
}

ngOnDestroy() {
   this.productSubscription.unsubscribe();
}
```

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 25]
```diff
@@ -24,4 +27,13 @@ export class ProductDetailsPageComponent {
     title: 'Half Jacket + Skiny Trousers + Boot leather',
     description: 'Lorem Lorem Lorem',
   };
+  routeId: String;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: remove

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 26]
```diff
@@ -1,13 +1,16 @@
...
 
 @Component({
   selector: 'app-product-details-page',
   templateUrl: './product-details-page.component.html',
   styleUrls: ['./product-details-page.component.scss']
 })
-export class ProductDetailsPageComponent {
+export class ProductDetailsPageComponent implements OnInit {
 
   productInfo = {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: remove

it should be taken from product response

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 27]
```diff
@@ -24,4 +27,13 @@ export class ProductDetailsPageComponent {
     title: 'Half Jacket + Skiny Trousers + Boot leather',
     description: 'Lorem Lorem Lorem',
   };
+  routeId: String;
+  relatedProducts: IProduct[] = [];
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: remove

File: src/app/product-details-page/components/product-details-page/product-details-page.component.html [hunk 28]
```diff
@@ -1 +1,2 @@
 <app-product-description [product]="productInfo"></app-product-description>
+<app-related-products [products]="relatedProducts" >
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.html]: <app-related-products [products]="product.relatedProducts" >

File: src/app/core/services/product-details-page.service.ts [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductDetailsPageService {
+
+
+  constructor(private dataService: DataService) {}
+
+  getProduct(id: String) {
```
Comment (dverbovyi) [inline src/app/core/services/product-details-page.service.ts]: specify the return type
```ts
getProduct(id: String): Observable<IProduct> {
...
}
```

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/product-details-page/components/product-details-page/product-details-page.component.ts [hunk 30]
```diff
@@ -1,27 +1,30 @@
...
-    swatches: [{ 'color': 'red', 'imgSrc': '' }, { 'color': 'black', 'imgSrc': 'http://www.roasterydepartment.com/images/large/reebok/Beautiful%20Reebok%20Women%20Reebok%20Full%20Zip%20Fleece%20Jacket%20Women%20Reebok%20Jackets%20SW87_LRG.jpg' }, { 'color': 'grey', 'imgSrc': 'http://www.roasterydepartment.com/images/large/reebok/Popular%20Reebok%20Women%20Reebok%20Full%20Zip%20Fleece%20Jacket%20Women%20Reebok%20Jackets%20MJ85_LRG.jpg' }, { 'color': 'blue', 'imgSrc': 'http://www.roasterydepartment.com/images/large/reebok/Cheap%20Reebok%20Women%20Reebok%20Full%20Zip%20Fleece%20Jacket%20Women%20Reebok%20Jackets%20PE94_LRG.jpg' }],
-    availability: [ProductAvailabilityState.IN_STORE, ProductAvailabilityState.ONLINE_ONLY],
-    thumbnailImageSrc: 'http://therxreview.com/wp-content/uploads/2013/05/Reebok-Track-Jacket.png',
-    sizes: [ProductSize.S, ProductSize.M, ProductSize.L, ProductSize.XL],
-    addedToCart: false,
-    addedToWishList: true,
-    title: 'Half Jacket + Skiny Trousers + Boot leather',
-    description: 'Lorem Lorem Lorem',
-  };
+  productSubscription;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.ts]: private

File: src/app/product-details-page/components/product-details-page/product-details-page.component.html [hunk 31]
```diff
@@ -1 +1,2 @@
 <app-product-description [product]="productInfo"></app-product-description>
+<app-related-products [products]="product?.relatedProducts"></app-related-products>
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/product-details-page.component.html]: it's better not to instantiate related-product component at all if there are no products
```html
<app-related-products *ngIf="product?.relatedProducts" [products]="product.relatedProducts"></app-related-products>
```

File: src/app/core/services/product-details-page.service.spec.ts [hunk 32]
```diff
@@ -0,0 +1,27 @@
+import { TestBed } from '@angular/core/testing';
+
+import { ProductDetailsPageService } from './product-details-page.service';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { HttpClientModule } from '@angular/common/http';
```
Comment (dverbovyi) [inline src/app/core/services/product-details-page.service.spec.ts]: ### Breaking the Styleguide detected

Separete "**Theirs** first from **Ours**". Aplly this pattern for all files

#### Example how should be
https://angular.io/guide/styleguide#import-line-spacing

screen shot 2019-01-31 at 15 15 17

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.html [hunk 33]
```diff
@@ -0,0 +1,8 @@
+<div *ngIf="products?.length">
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-details-page/related-products/related-products.component.html]: Move this check to the parent component and just don't render the child one. Child component should definitely have the product and doesn't validate it's absence

File: src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts (hunk 23 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-details-page/related-products/related-products.component.ts]: not resolved

File: src/app/shared/components/advertising-area/advertising-area.component.spec.ts [hunk 34]
```diff
@@ -1,8 +1,8 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { AdvertisingAreaComponent } from './advertising-area.component';
-import { SafePipe } from 'src/app/pipes/safe.pipe';
 import { By } from '@angular/platform-browser';
```
Comment (dverbovyi) [inline src/app/shared/components/advertising-area/advertising-area.component.spec.ts]: broken styleguide

own imports shoud come after core and third-parties

**Review by dverbovyi — CHANGES_REQUESTED** [review]

**Review by dverbovyi — APPROVED** [review]

**Pull request merged**